@fwyzard I guess the latest update was just a rebase, as it is the code does not seem functional yet:

```
13:32 cmsdev25 7647> cmsDriver.py step3  --conditions auto:phase1_2018_realistic -n 10 --era Run2_2018 --eventcontent RECOSIM,MINIAODSIM,DQM --runUnscheduled  -s RAW2DIGI,L1Reco,RECO,RECOSIM,EI,PAT,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@ExtraHLT+@miniAODDQM --datatier GEN-SIM-RECO,MINIAODSIM,DQMIO --geometry DB:Extended --filein  file:step2.root  --fileout file:step3.root --no_exec
RAW2DIGI,L1Reco,RECO,RECOSIM,EI,PAT,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@ExtraHLT+@miniAODDQM
We have determined that this is simulation (if not, rerun cmsDriver.py with --data)
with DB:
entry file:step2.root
Step: RAW2DIGI Spec: 
Step: L1Reco Spec: 
Step: RECO Spec: 
Step: RECOSIM Spec: 
Step: EI Spec: 
Step: PAT Spec: 
Step: VALIDATION Spec: ['@standardValidation', '@miniAODValidation']
@standardValidation+@miniAODValidation in preparing validation
Step: DQM Spec: ['@standardDQM', '@ExtraHLT', '@miniAODDQM']
Traceback (most recent call last):
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/bin/slc7_amd64_gcc820/cmsDriver.py", line 56, in <module>
    run()
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/bin/slc7_amd64_gcc820/cmsDriver.py", line 28, in run
    configBuilder.prepare()
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/python/Configuration/Applications/ConfigBuilder.py", line 2188, in prepare
    self.pythonCfgCode += dumpPython(self.process,path)
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/python/Configuration/Applications/ConfigBuilder.py", line 95, in dumpPython
    return "process."+name+" = " + theObject.dumpPython("process")
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/python/FWCore/ParameterSet/SequenceTypes.py", line 280, in dumpPython
    s = self.dumpPythonNoNewline(options)
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/python/FWCore/ParameterSet/SequenceTypes.py", line 286, in dumpPythonNoNewline
    s =self._seq.dumpSequencePython(options)
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/python/FWCore/ParameterSet/SequenceTypes.py", line 168, in dumpSequencePython
    itemDump = item.dumpSequencePython(options)
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/python/FWCore/ParameterSet/SequenceTypes.py", line 305, in dumpSequencePython
    return _Labelable.dumpSequencePython(self, options)
  File "/build/fabiocos/110X/27305/CMSSW_11_0_X_2019-10-10-2300/python/FWCore/ParameterSet/Mixins.py", line 541, in dumpSequencePython
    if options.isCfg:
AttributeError: 'str' object has no attribute 'isCfg'
```

--- a/FWCore/ParameterSet/python/Config.py
+++ b/FWCore/ParameterSet/python/Config.py
@@ -713,11 +713,13 @@ class Process(object):
         for name,item in items:
             returnValue +=options.indentation()+typeName+' '+name+' = '+item.dumpConfig(options)
         return returnValue
+
     def _dumpConfigUnnamedList(self,items,typeName,options):
         returnValue = ''
         for name,item in items:
             returnValue +=options.indentation()+typeName+' = '+item.dumpConfig(options)
         return returnValue
+
     def _dumpConfigOptionallyNamedList(self,items,typeName,options):
         returnValue = ''
         for name,item in items:
@@ -725,6 +727,7 @@ class Process(object):
                 name = ''
             returnValue +=options.indentation()+typeName+' '+name+' = '+item.dumpConfig(options)
         return returnValue
+
     def dumpConfig(self, options=PrintOptions()):
         """return a string containing the equivalent process defined using the old configuration language"""
         config = "process "+self.__name+" = {\n"
@@ -790,16 +793,19 @@ class Process(object):
         config += "}\n"
         options.unindent()
         return config
+
     def _dumpConfigESPrefers(self, options):
         result = ''
         for item in six.itervalues(self.es_prefers_()):
             result +=options.indentation()+'es_prefer '+item.targetLabel_()+' = '+item.dumpConfig(options)
         return result
+
     def _dumpPythonSubProcesses(self, l, options):
         returnValue = ''
         for item in l:
             returnValue += item.dumpPython(options)+'\n\n'
         return returnValue
+
     def _dumpPythonList(self, d, options):
         returnValue = ''
         if isinstance(d, DictTypes.SortedKeysDict):
@@ -809,6 +815,17 @@ class Process(object):
             for name,item in sorted(d.items()):
                 returnValue +='process.'+name+' = '+item.dumpPython(options)+'\n\n'
         return returnValue
+
+    def _splitPythonList(self, d, options):
+        parts = DictTypes.SortedKeysDict()
+        for name, item in d.items() if isinstance(d, DictTypes.SortedKeysDict) else sorted(d.items()):
+            dependencies = item.directDependencies()
+            parts[name] = '\n'.join('from ' + module + '_cfi import *' for module in dependencies)
+            if dependencies:
+              parts[name] += '\n\n'
+            parts[name] += name + ' = ' + item.dumpPython(options)
+        return parts
+
     def _validateSequence(self, sequence, label):
         # See if every module has been inserted into the process
         try:
@@ -817,6 +834,7 @@ class Process(object):
             sequence.visit(visitor)
         except:
             raise RuntimeError("An entry in sequence "+label + ' has no label')
+
     def _validateTask(self, task, label):
         # See if every module and service has been inserted into the process
         try:
@@ -825,6 +843,7 @@ class Process(object):
             task.visit(visitor)
         except:
             raise RuntimeError("An entry in task " + label + ' has not been attached to the process')
+
     def _itemsInDependencyOrder(self, processDictionaryOfItems):
         # The items can be Sequences or Tasks and the input
         # argument should either be the dictionary of sequences
@@ -881,11 +900,19 @@ class Process(object):
                         while deps2.count(label):
                             deps2.remove(label)
         return returnValue
+
     def _dumpPython(self, d, options):
         result = ''
         for name, value in sorted(six.iteritems(d)):
             result += value.dumpPythonAs(name,options)+'\n'
         return result
+
+    def _splitPython(self, d, options):
+        result = {}
+        for name, value in sorted(six.iteritems(d)):
+            result[name] = value.dumpPythonAs(name,options)+'\n'
+        return result
+
     def dumpPython(self, options=PrintOptions()):
         """return a string containing the equivalent process defined using python"""
         specialImportRegistry._reset()
@@ -919,6 +946,60 @@ class Process(object):
             header += "\n" + "\n".join(imports)
         header += "\n\n"
         return header+result
+
+    def splitPython(self, options=PrintOptions()):
+        """return a map of names to python configuration fragments"""
+        specialImportRegistry._reset()
+        # extract individual fragments
+        options.isCfg = False
+        header = "import FWCore.ParameterSet.Config as cms"
+        result = ''
+        parts = {}
+
+        result = 'process = cms.Process("' + self.__name + '")\n\n'
+
+        if self.source_():
+            parts['source'] = 'source = ' + self.source_().dumpPython(options)
+
+        if self.looper_():
+            parts['looper'] = 'looper = ' + self.looper_().dumpPython()
+
+        parts.update(self._splitPythonList(self.psets, options))
+        parts.update(self._splitPythonList(self.vpsets, options))
+        # FIXME
+        #parts.update(self._splitPythonSubProcesses(self.subProcesses_(), options))
+        if len(self.subProcesses_()):
+          sys.stderr.write("error: subprocesses are not supported yet\n\n")
+        parts.update(self._splitPythonList(self.producers_(), options))
+        parts.update(self._splitPythonList(self.switchProducers_(), options))
+        parts.update(self._splitPythonList(self.filters_() , options))
+        parts.update(self._splitPythonList(self.analyzers_(), options))
+        parts.update(self._splitPythonList(self.outputModules_(), options))
+        parts.update(self._splitPythonList(self.services_(), options))
+        parts.update(self._splitPythonList(self.es_producers_(), options))
+        parts.update(self._splitPythonList(self.es_sources_(), options))
+        parts.update(self._splitPython(self.es_prefers_(), options))
+        parts.update(self._splitPythonList(self._itemsInDependencyOrder(self.tasks), options))
+        parts.update(self._splitPythonList(self._itemsInDependencyOrder(self.sequences), options))
+        parts.update(self._splitPythonList(self.paths_(), options))
+        parts.update(self._splitPythonList(self.endpaths_(), options))
+        parts.update(self._splitPythonList(self.aliases_(), options))
+
+        for name in parts:
+            result += 'process.load("' + name + '_cfi")\n'
+            with open(name + '_cfi.py', 'w') as f:
+              f.write(header + '\n\n')
+              f.write(parts[name])
+
+        if not self.schedule_() == None:
+            result += 'process.schedule = ' + self.schedule.dumpPython(options)
+
+        imports = specialImportRegistry.getSpecialImports()
+        if len(imports) > 0:
+            header += "\n" + "\n".join(imports)
+        header += "\n\n"
+        return header+result
+
     def _replaceInSequences(self, label, new):
         old = getattr(self,label)
         #TODO - replace by iterator concatenation
@@ -1328,7 +1409,7 @@ class SubProcess(_Unlabelable):
         self.__process = process
         self.__SelectEvents = SelectEvents
         self.__outputCommands = outputCommands
-    def dumpPython(self,options=PrintOptions()):
+    def dumpPython(self, options=PrintOptions()):
         out = "parentProcess"+str(hash(self))+" = process\n"
         out += self.__process.dumpPython()
         out += "childProcess = process\n"

--- a/FWCore/ParameterSet/python/Config.py
+++ b/FWCore/ParameterSet/python/Config.py
@@ -4,6 +4,7 @@
 from __future__ import print_function
 from __future__ import absolute_import
 import six
+import os
 from  .Options import Options
 options = Options()
 
@@ -816,14 +817,22 @@ class Process(object):
                 returnValue +='process.'+name+' = '+item.dumpPython(options)+'\n\n'
         return returnValue
 
-    def _splitPythonList(self, d, options):
+    def _splitPythonList(self, subfolder, d, options):
         parts = DictTypes.SortedKeysDict()
         for name, item in d.items() if isinstance(d, DictTypes.SortedKeysDict) else sorted(d.items()):
+            code = ''
             dependencies = item.directDependencies()
-            parts[name] = '\n'.join('from ' + module + '_cfi import *' for module in dependencies)
+            for module_subfolder, module in dependencies:
+                module = module + '_cfi'
+                if options.useSubdirectories and module_subfolder:
+                    module = module_subfolder + '.' + module
+                if options.targetDirectory is not None:
+                    module = options.targetDirectory + '.' + module
+                code += 'from ' + module + ' import *\n'
             if dependencies:
-              parts[name] += '\n\n'
-            parts[name] += name + ' = ' + item.dumpPython(options)
+                code += '\n'
+            code += name + ' = ' + item.dumpPython(options)
+            parts[name] = subfolder, code
         return parts
 
     def _validateSequence(self, sequence, label):
@@ -907,10 +916,10 @@ class Process(object):
             result += value.dumpPythonAs(name,options)+'\n'
         return result
 
-    def _splitPython(self, d, options):
+    def _splitPython(self, subfolder, d, options):
         result = {}
         for name, value in sorted(six.iteritems(d)):
-            result[name] = value.dumpPythonAs(name,options)+'\n'
+            result[name] = subfolder, value.dumpPythonAs(name, options) + '\n'
         return result
 
     def dumpPython(self, options=PrintOptions()):
@@ -947,7 +956,7 @@ class Process(object):
         header += "\n\n"
         return header+result
 
-    def splitPython(self, options=PrintOptions()):
+    def splitPython(self, options = PrintOptions()):
         """return a map of names to python configuration fragments"""
         specialImportRegistry._reset()
         # extract individual fragments
@@ -959,39 +968,57 @@ class Process(object):
         result = 'process = cms.Process("' + self.__name + '")\n\n'
 
         if self.source_():
-            parts['source'] = 'source = ' + self.source_().dumpPython(options)
+            parts['source'] = None, 'source = ' + self.source_().dumpPython(options)
 
         if self.looper_():
-            parts['looper'] = 'looper = ' + self.looper_().dumpPython()
+            parts['looper'] = None, 'looper = ' + self.looper_().dumpPython()
 
-        parts.update(self._splitPythonList(self.psets, options))
-        parts.update(self._splitPythonList(self.vpsets, options))
+        parts.update(self._splitPythonList('psets', self.psets, options))
+        parts.update(self._splitPythonList('psets', self.vpsets, options))
         # FIXME
         #parts.update(self._splitPythonSubProcesses(self.subProcesses_(), options))
         if len(self.subProcesses_()):
           sys.stderr.write("error: subprocesses are not supported yet\n\n")
-        parts.update(self._splitPythonList(self.producers_(), options))
-        parts.update(self._splitPythonList(self.switchProducers_(), options))
-        parts.update(self._splitPythonList(self.filters_() , options))
-        parts.update(self._splitPythonList(self.analyzers_(), options))
-        parts.update(self._splitPythonList(self.outputModules_(), options))
-        parts.update(self._splitPythonList(self.services_(), options))
-        parts.update(self._splitPythonList(self.es_producers_(), options))
-        parts.update(self._splitPythonList(self.es_sources_(), options))
-        parts.update(self._splitPython(self.es_prefers_(), options))
-        parts.update(self._splitPythonList(self._itemsInDependencyOrder(self.tasks), options))
-        parts.update(self._splitPythonList(self._itemsInDependencyOrder(self.sequences), options))
-        parts.update(self._splitPythonList(self.paths_(), options))
-        parts.update(self._splitPythonList(self.endpaths_(), options))
-        parts.update(self._splitPythonList(self.aliases_(), options))
+        parts.update(self._splitPythonList('modules', self.producers_(), options))
+        parts.update(self._splitPythonList('modules', self.switchProducers_(), options))
+        parts.update(self._splitPythonList('modules', self.filters_() , options))
+        parts.update(self._splitPythonList('modules', self.analyzers_(), options))
+        parts.update(self._splitPythonList('modules', self.outputModules_(), options))
+        parts.update(self._splitPythonList('services', self.services_(), options))
+        parts.update(self._splitPythonList('eventsetup', self.es_producers_(), options))
+        parts.update(self._splitPythonList('eventsetup', self.es_sources_(), options))
+        parts.update(self._splitPython('eventsetup', self.es_prefers_(), options))
+        parts.update(self._splitPythonList('tasks', self._itemsInDependencyOrder(self.tasks), options))
+        parts.update(self._splitPythonList('sequences', self._itemsInDependencyOrder(self.sequences), options))
+        parts.update(self._splitPythonList('paths', self.paths_(), options))
+        parts.update(self._splitPythonList('paths', self.endpaths_(), options))
+        parts.update(self._splitPythonList('modules', self.aliases_(), options))
 
-        for name in parts:
-            result += 'process.load("' + name + '_cfi")\n'
-            with open(name + '_cfi.py', 'w') as f:
+        if options.targetDirectory is not None:
+            if not os.path.isdir(options.targetDirectory):
+                os.makedirs(options.targetDirectory)
+            open(options.targetDirectory + '/__init__.py', 'w').close()
+
+        if options.useSubdirectories:
+          for sub in 'psets', 'modules', 'services', 'eventsetup', 'tasks', 'sequences', 'paths':
+            if options.targetDirectory is not None:
+                sub = options.targetDirectory + '/' + sub
+            if not os.path.isdir(sub):
+                os.makedirs(sub)
+            open(sub + '/__init__.py', 'w').close()
+
+        for (name, (subfolder, code)) in six.iteritems(parts):
+            filename = name + '_cfi'
+            if options.useSubdirectories and subfolder:
+              filename = subfolder + '/' + filename
+            if options.targetDirectory is not None:
+              filename = options.targetDirectory + '/' + filename
+            result += 'process.load("%s")\n' % filename
+            with open('%s.py' % filename, 'w') as f:
               f.write(header + '\n\n')
-              f.write(parts[name])
+              f.write(code)
 
-        if not self.schedule_() == None:
+        if self.schedule_() is not None:
             result += 'process.schedule = ' + self.schedule.dumpPython(options)
 
         imports = specialImportRegistry.getSpecialImports()

--- a/FWCore/ParameterSet/python/Mixins.py
+++ b/FWCore/ParameterSet/python/Mixins.py
@@ -9,10 +9,12 @@ class _ConfigureComponent(object):
         return False
 
 class PrintOptions(object):
-    def __init__(self, indent = 0, deltaIndent = 4, process = True):
+    def __init__(self, indent = 0, deltaIndent = 4, process = True, targetDirectory = None, useSubdirectories = False):
         self.indent_= indent
         self.deltaIndent_ = deltaIndent
         self.isCfg = process
+        self.targetDirectory = targetDirectory
+        self.useSubdirectories = useSubdirectories
     def indentation(self):
         return ' '*self.indent_
     def indent(self):

--- a/FWCore/ParameterSet/python/Mixins.py
+++ b/FWCore/ParameterSet/python/Mixins.py
@@ -452,6 +452,9 @@ class _TypedParameterizable(_Parameterizable):
                             params[name] = getattr(default,name)
                         return params
         return None
+
+    def directDependencies(self):
+        return []
     
     def dumpConfig(self, options=PrintOptions()):
         config = self.__type +' { \n'
@@ -534,7 +537,10 @@ class _Labelable(object):
     def dumpSequenceConfig(self):
         return str(self.__label)
     def dumpSequencePython(self, options=PrintOptions()):
-        return 'process.'+str(self.__label)
+        if options.isCfg:
+            return 'process.'+str(self.__label)
+        else:
+            return str(self.__label)
     def _findDependencies(self,knownDeps,presentDeps):
         #print 'in labelled'
         myDeps=knownDeps.get(self.label_(),None)

--- a/FWCore/ParameterSet/python/Modules.py
+++ b/FWCore/ParameterSet/python/Modules.py
@@ -359,6 +359,10 @@ class SwitchProducer(EDProducer):
         result += "\n)\n"
         return result
 
+    def directDependencies(self):
+        # XXX FIXME handle SwitchProducer dependencies
+        return []
+
     def nameInProcessDesc_(self, myname):
         return myname
     def moduleLabel_(self, myname):

--- a/FWCore/ParameterSet/python/SequenceTypes.py
+++ b/FWCore/ParameterSet/python/SequenceTypes.py
@@ -160,6 +160,7 @@ class _SequenceCollection(_Sequenceable):
         return returnValue
     def _appendToCollection(self,collection):
         collection.extend(self._collection)
+
     def dumpSequencePython(self, options=PrintOptions()):
         returnValue = ''
         separator = ''
@@ -169,19 +170,30 @@ class _SequenceCollection(_Sequenceable):
                 returnValue += (separator + itemDump)
                 separator = '+'
         return returnValue
+
     def dumpSequenceConfig(self):
         returnValue = self._collection[0].dumpSequenceConfig()
         for m in self._collection[1:]:
             returnValue += '&'+m.dumpSequenceConfig()        
         return returnValue
+
+    def directDependencies(self):
+        return [ item.label() for item in self._collection ]
+        #options = PrintOptions()
+        #options.isCfg = False
+        #return [ name for name in (item.dumpSequencePython(options) for item in self._collection) if name ]
+
     def visitNode(self,visitor):
         for m in self._collection:
             m.visitNode(visitor)
+
     def resolve(self, processDict,keepIfCannotResolve=False):
         self._collection = [x.resolve(processDict,keepIfCannotResolve) for x in self._collection]
         return self
+
     def index(self,item):
         return self._collection.index(item)
+
     def insert(self,index,item):
         self._collection.insert(index,item)
     def _replaceIfHeldDirectly(self,original,replacement):
@@ -256,15 +268,18 @@ class _ModuleSequenceType(_ConfigureComponent, _Labelable):
         v = ExpandVisitor(type(self))
         self.visit(v)
         return v.resultString()
+
     def dumpConfig(self, options):
         s = ''
         if self._seq is not None:
             s = self._seq.dumpSequenceConfig()
         return '{'+s+'}\n'
+
     def dumpPython(self, options=PrintOptions()):
         """Returns a string which is the python representation of the object"""
         s = self.dumpPythonNoNewline(options)
         return s + "\n"
+
     def dumpPythonNoNewline(self, options=PrintOptions()):
         s=''
         if self._seq is not None:
@@ -282,6 +297,7 @@ class _ModuleSequenceType(_ConfigureComponent, _Labelable):
         if len(associationContents) > 254:
             return 'cms.'+type(self).__name__+'(*['+s+'])'
         return 'cms.'+type(self).__name__+'('+s+')'
+
     def dumpSequencePython(self, options=PrintOptions()):
         """Returns a string which contains the python representation of just the internal sequence"""
         # only dump the label, if possible
@@ -295,6 +311,7 @@ class _ModuleSequenceType(_ConfigureComponent, _Labelable):
               return '('+s+')'
             return ''
         return self.dumpPythonNoNewline(options)
+
     def dumpSequenceConfig(self):
         """Returns a string which contains the old config language representation of just the internal sequence"""
         # only dump the label, if possible
@@ -305,21 +322,34 @@ class _ModuleSequenceType(_ConfigureComponent, _Labelable):
             if self._seq is None:
                 return ''
             return '('+self._seq.dumpSequenceConfig()+')'
+
     def __repr__(self):
         s = ''
         if self._seq is not None:
            s = str(self._seq)
         return "cms."+type(self).__name__+'('+s+')\n'
+
+    def directDependencies(self):
+        """Returns the list of modules and other entities that are directly used"""
+        result = []
+        if self._seq:
+          result += self._seq.directDependencies()
+        if self._tasks:
+          result += [ task.label() for task in self._tasks ]
+        return result
+
     def moduleNames(self):
         """Returns a set containing the names of all modules being used"""
         result = set()
         visitor = NodeNameVisitor(result)
         self.visit(visitor)
         return result
+
     def contains(self, mod):
         visitor = ContainsModuleVisitor(mod)
         self.visit(visitor)
         return visitor.result()
+
     def copy(self):
         returnValue =_ModuleSequenceType.__new__(type(self))
         if self._seq is not None:
@@ -328,6 +358,7 @@ class _ModuleSequenceType(_ConfigureComponent, _Labelable):
             returnValue.__init__()
         returnValue._tasks = OrderedSet(self._tasks)
         return returnValue
+
     def copyAndExclude(self,listOfModulesToExclude):
         """Returns a copy of the sequence which excludes those module in 'listOfModulesToExclude'"""
         # You can exclude instances of these types EDProducer, EDFilter, OutputModule,

--- a/FWCore/ParameterSet/python/SequenceTypes.py
+++ b/FWCore/ParameterSet/python/SequenceTypes.py
@@ -178,10 +178,18 @@ class _SequenceCollection(_Sequenceable):
         return returnValue
 
     def directDependencies(self):
-        return [ item.label() for item in self._collection ]
-        #options = PrintOptions()
-        #options.isCfg = False
-        #return [ name for name in (item.dumpSequencePython(options) for item in self._collection) if name ]
+        dependencies = []
+        for item in self._collection:
+            if type(item).__name__ in ("EDFilter", "EDProducer", "EDAnalyzer", "OutputModule"):
+                t = 'modules'
+            elif type(item).__name__ in ("_SequenceIgnore", "_SequenceNegation"):
+                t = 'modules'
+            elif type(item).__name__ in ("Sequence"):
+                t = 'sequences'
+            else:
+                t = ''
+            dependencies.append((t, item.label()))
+        return dependencies
 
     def visitNode(self,visitor):
         for m in self._collection:
@@ -335,7 +343,7 @@ class _ModuleSequenceType(_ConfigureComponent, _Labelable):
         if self._seq:
           result += self._seq.directDependencies()
         if self._tasks:
-          result += [ task.label() for task in self._tasks ]
+          result += [ ('tasks', task.label()) for task in self._tasks ]
         return result
 
     def moduleNames(self):

--- a/FWCore/ParameterSet/python/Types.py
+++ b/FWCore/ParameterSet/python/Types.py
@@ -827,6 +827,9 @@ class PSet(_ParameterTypeBase,_Parameterizable,_ConfigureComponent,_Labelable):
         return config
     def dumpPython(self, options=PrintOptions()):
         return self.pythonTypeName()+"(\n"+_Parameterizable.dumpPython(self, options)+options.indentation()+")"
+    # XXX FIXME handle refToPSet
+    def directDependencies(self):
+        return []
     def clone(self, **params):
         myparams = self.parameters_()
         _modifyParametersFromDict(myparams, params, self._Parameterizable__raiseBadSetAttr)

--- a/FWCore/ParameterSet/python/Types.py
+++ b/FWCore/ParameterSet/python/Types.py
@@ -1356,6 +1356,10 @@ class EDAlias(_ConfigureComponent,_Labelable,_Parameterizable):
             options.unindent()
         return '\n'.join(resultList)+'\n)'
 
+    # an EDAlias only references other modules by label, so it does not need their definition
+    def directDependencies(self):
+        return []
+
 if __name__ == "__main__":
 
     import unittest

--- a/FWCore/ParameterSet/scripts/edmConfigSplit
+++ b/FWCore/ParameterSet/scripts/edmConfigSplit
@@ -1,0 +1,26 @@
+#! /usr/bin/env python
+
+from optparse import OptionParser
+import sys
+import os
+import imp
+
+# make the behaviour of 'cmsRun file.py' and 'edmConfigDump file.py' more consistent
+sys.path.append(os.getcwd())
+
+parser = OptionParser()
+parser.usage = "%prog <file> : expand this python configuration"
+
+(options,args) = parser.parse_args()
+
+if len(args)!=1:
+    parser.print_help()
+    sys.exit(1)
+
+filename = args[0]
+handle = open(filename, 'r')
+cfo = imp.load_source("pycfg", filename, handle)
+cmsProcess = cfo.process
+handle.close()
+
+print cmsProcess.splitPython()

--- a/FWCore/ParameterSet/scripts/edmConfigSplit
+++ b/FWCore/ParameterSet/scripts/edmConfigSplit
@@ -1,26 +1,55 @@
 #! /usr/bin/env python
 
-from optparse import OptionParser
 import sys
 import os
 import imp
+import argparse
 
-# make the behaviour of 'cmsRun file.py' and 'edmConfigDump file.py' more consistent
+from FWCore.ParameterSet.Mixins import PrintOptions
+
+# customise the HelpFormatter to show additional parameters in the usage line
+class PartialHelpFormatter(argparse.HelpFormatter):
+  def _format_usage(self, usage, actions, groups, prefix):
+    partial = super(PartialHelpFormatter, self)._format_usage(usage, actions, groups, prefix).strip()
+    return partial + ' [--] [OPTIONS ...]\n\n'
+
+
+parser = argparse.ArgumentParser(formatter_class = PartialHelpFormatter,
+  description = '%(prog)s splits the given CMSSW configuration file into one file per top-level object, and outputs the configuration for the main process object.',
+  epilog = 'Additional options and arguments are passed to the configuration file unchanged (e.g. for use by VarParsing.py).')
+parser.add_argument('file', metavar = 'FILE', type = argparse.FileType('r'))
+
+parser.add_argument('-o', '--output',
+  metavar = 'OUT',
+  type = argparse.FileType('w'),
+  default = sys.stdout,
+  help = 'write the process configuration to %(metavar)s instead of standard output')
+parser.add_argument('-d', '--output-directory',
+  metavar = 'DIR',
+  type = str,
+  default = None,
+  help = 'create the individual files and subdirectories under %(metavar)s instead of the current directory; if %(metavar)s does not exist it will be created first')
+parser.add_argument('-s', '--subdirectories',
+  action = "store_true",
+  default = False,
+  help = 'create subdirectories for different modules categories')
+
+args, other = parser.parse_known_args()
+
+# delete all arguments, so they are not "seen" by the confguration file being split
+del sys.argv[1:]
+sys.argv.append(args.file.name)
+sys.argv.extend(other)
+
+# make the behaviour of 'cmsRun file.py' and 'edmConfigSplit file.py' more consistent
 sys.path.append(os.getcwd())
 
-parser = OptionParser()
-parser.usage = "%prog <file> : expand this python configuration"
+cfo = imp.load_source("pycfg", args.file.name, args.file)
+process = cfo.process
+args.file.close()
 
-(options,args) = parser.parse_args()
-
-if len(args)!=1:
-    parser.print_help()
-    sys.exit(1)
-
-filename = args[0]
-handle = open(filename, 'r')
-cfo = imp.load_source("pycfg", filename, handle)
-cmsProcess = cfo.process
-handle.close()
-
-print cmsProcess.splitPython()
+options = PrintOptions()
+options.useSubdirectories = args.subdirectories
+options.targetDirectory = args.output_directory
+args.output.write(process.splitPython(options))
+args.output.close()

